### PR TITLE
[inductor] add cat_linear_fused pre-grad pass for F.linear(cat(...))

### PR DESCRIPTION
**Pull request by reger-men**
Pre-grad FX pass that rewrites `F.linear(torch.cat([t_0, ...], dim=-1), W, b)` into a reduce-sum of per-piece `F.linear` calls on contiguous weight slices. Avoids materialising the cat on the forward path (and the cat's grad on the backward path), which on bf16 GEMMs is a measurable HBM-bandwidth win.

Conservative gating: only fires when every cat operand is a last-dim slice with the same leading shape, bias is None (or only on the first partial linear), total cat width is below `MAX_TOTAL_CAT_WIDTH`, and the cat axis is the last axis (handles negative indexing).

Off by default; opt in via
```python
torch._inductor.config.pre_grad_custom_pass = cat_linear_fused_pre_grad_pass
```

Implementation note: the helper `_val_of` uses an explicit `is None` check rather than `or`, because `node.meta.get("val") or node.meta.get("example_value")` calls `Tensor.__bool__` and raises when `meta["val"]` is a multi-element tensor (which can happen on a post-grad graph or a test graph that populates `meta["val"]` directly).

Test plan
- `test_cat_linear_fused.py` matcher tests: canonical pattern fires, 3-part fires, rejects more than MAX_PARTS, rejects pieces below MIN_PIECE_WIDTH, rejects total > MAX_TOTAL_CAT_WIDTH, rejects `mul`-parented operand, rejects cat on a non-last axis
- integration tests: `torch.compile` with the pass installed produces results within 1e-4 of eager and bumps the `cat_linear_fused` counter; with the pass not installed the counter stays at 0

### CODE REVIEW

**Comment by pruthvistony** [comment]
Better to raise this PR into pytorch/pytorch upstream and properly tested in CI.
Develop branch on rocm/pytorch is a copy of upstream, so we dont want local changes.

**Comment by jeffdaily** [comment]
Initial upstream PR https://github.com/pytorch/pytorch/pull/186641.  Replaced by https://github.com/pytorch/pytorch/pull/187074 but was messed up by force merge.  Finally landed upstream as https://github.com/pytorch/pytorch/pull/187880.
